src/maths/modulo.rs

```diff
@@ -1,25 +1,59 @@
-type Data = i64;
+type Mod = u128;
 
-const MOD: Data = 1e9 as Data + 7;
+const MOD: Mod = 1e9 as Mod + 7;
 
-pub trait Modulo {
-    fn add(&mut self, other: Data);
-
-    fn sub(&mut self, other: Data);
+pub trait ModMath {
+    fn madd(&self, other: Mod) -> Mod;
+    fn msub(&self, other: Mod) -> Mod;
+    fn mmul(&self, other: Mod) -> Mod;
+    fn mdiv(&self, other: Mod) -> Mod;
+    fn mpow(&self, power: Mod) -> Mod;
+    fn minverse(&self) -> Mod;
 }
 
-impl Modulo for Data {
-    fn add(&mut self, other: Data) {
-        *self += other;
-        if *self >= MOD {
-            *self -= MOD;
+impl ModMath for Mod {
+    fn madd(&self, other: Mod) -> Mod {
+        ((self % MOD) + (other % MOD)) % MOD
```

Comment:
so, the thing with add compare and sub, rather than add and modulo is the fact the the previous one is faster. And if the number of operations are a lot, then the time difference is huge. I observed a difference of 1.5 sec on CF between a problem. lets try and keep that standard. 

The idea is, if something is under mods limit and all operation are performed using the associated methods, then the value will always be withing mods limit. [valid for all operations]

Reply:
Ah understood. Do you still want to mutate `self` or return the answer?

Reply:
we can do both. that way we can chain operations. like x.add(a).add(b).mul(c).inv()

Reply:
We can chain operations even if we don't mutate `self`. `x.madd(a).msub(b)` applies `msub()` to the result of `x.madd(a)`.

(I wrote the test that way to show chaining)

Reply:
yeh, but it needs one more assign, like x = x.madd(a).msub(b); 
with cp related algos, we have as much as we can, so we save one allocation here where otherwise it would need to allocate a variable for storing temp data.

my mutating self, we could just say (its kind of like pushing values in an aggregator function)
x.madd(a).madd(b);
and we are done.